File: typescipt/src/language/expressions/functions/newFunction.ts


export class NewFunction extends ExpressionFunction, IHasNodeDependencies {
   public const string Name = `new`;

   protected string FunctionHelp => $`{Name} expects 1 argument (Function.Parameters)`;

   public MemberAccessLiteral TypeLiteral

   public Expression ValueExpression

   public ComplexTypeReference Type { get; private set; }

   constructor(valueExpression: Expression, reference: SourceReference)
     {
     super(reference);
     ValueExpression = valueExpression ?? throw new Error(nameof(valueExpression));
     TypeLiteral = (valueExpression as MemberAccessExpression)?.MemberAccessLiteral;
   }

   public getDependencies(rootNodeList: RootNodeList): Array<IRootNode> {
     if (Type != null) yield return rootNodeList.GetNode(Type.Name);
   }

   public static create(reference: SourceReference, expression: Expression): ExpressionFunction {
     return new NewFunction(expression, reference);
   }

   public override getChildren(): Array<INode> {
     yield return ValueExpression;
   }

   protected override validate(context: IValidationContext): void {
     let valueType = ValueExpression.deriveType(context);
     if (!(valueType is ComplexTypeReference complexTypeReference)) {
       context.logger.fail(this.reference,
         $`Invalid argument 1 'Value' should be of type 'ComplexTypeType' but is 'ValueType'. {FunctionHelp}`);
       return;
     }

     Type = complexTypeReference;
   }

   public override deriveReturnType(context: IValidationContext): VariableType {
     let nodeType = context.RootNodes.GetType(TypeLiteral.Parent);
     let typeReference = nodeType?.MemberType(TypeLiteral.Member, context) as ComplexTypeReference;
     return typeReference?.GetComplexType(context);
   }
}

File: src/Lexy.Compiler/Language/Expressions/Functions/SystemFunctions/NewFunction.cs
using System;
using System.Collections.Generic;
using Lexy.Compiler.Language.VariableTypes;
using Lexy.Compiler.Parser;
using Lexy.Compiler.Parser.Tokens;

namespace Lexy.Compiler.Language.Expressions.Functions.SystemFunctions;

public class NewFunction : FunctionCallExpression, IHasNodeDependencies
{
    public const string Name = "new";

    protected string FunctionHelp => $"{Name} expects 1 argument new(Function.Parameters)";

    public MemberAccessLiteralToken TypeLiteralToken { get; }

    public Expression ValueExpression { get; }

    public GeneratedType Type { get; private set; }

    private NewFunction(Expression valueExpression, ExpressionSource source)
        : base(source)
    {
        ValueExpression = valueExpression ?? throw new ArgumentNullException(nameof(valueExpression));
        TypeLiteralToken = (valueExpression as MemberAccessExpression)?.MemberAccessLiteralToken;
    }

    public IEnumerable<IComponentNode> GetDependencies(IComponentNodeList componentNodes)
    {
        if (Type != null) yield return componentNodes.GetNode(Type.Name);
    }

    public static FunctionCallExpression Create(ExpressionSource source, Expression expression)
    {
        return new NewFunction(expression, source);
    }

    public override IEnumerable<INode> GetChildren()
    {
        yield return ValueExpression;
    }

    protected override void Validate(IValidationContext context)
    {
        var valueType = ValueExpression.DeriveType(context);
        if (valueType is not GeneratedType generatedType)
        {
            context.Logger.Fail(Reference,
                $"Invalid argument 1. 'Value' should be of type 'GeneratedType' but is '{valueType?.GetType()}'. {FunctionHelp}");
            return;
        }

        Type = generatedType;
    }

    public override VariableType DeriveType(IValidationContext context)
    {
        var nodeType = context.ComponentNodes.GetType(TypeLiteralToken.Parent);
        return nodeType?.MemberType(TypeLiteralToken.Member, context.ComponentNodes) as GeneratedType;
    }
}